### PR TITLE
Feature/jel 329 modularize login system

### LINKED ISSUE #329: Move "internal login" system to be a module
Move "internal login" system to be a module using pre/post hooks.

### PR DESCRIPTION
**Pull request by OBCENEIKON**
Closes #329

### CODE REVIEW

File: spec/factories/authentications.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+# Indexes
+#
+#  index_authentications_on_provider_and_uid  (provider,uid)
+#  index_authentications_on_staff_id          (staff_id)
+#
+
+FactoryGirl.define do
+  factory :authentications do
+    provider 'facebook'
+    uid '1234567'
```
Comment (chriskacerguis) [inline spec/factories/authentications.rb]: Uhhh...should that "uid" be there???

Comment (chriskacerguis) [inline spec/factories/authentications.rb]: @OBCENEIKON ???

File: Gemfile [hunk 2]
```diff
@@ -36,6 +36,8 @@ gem 'responders'
 # Authentication
 gem 'devise'
 gem 'ruby-saml'
+gem 'omniauth'
+gem 'omniauth-ldap'
```
Comment (calebthompson) [inline Gemfile]: Do we want this built in? Should this instead be a module we can add to show off how others could do this, as we did with jellyfish_fog_aws and jellyfish-azure?

Comment (chriskacerguis) [inline Gemfile]: @OBCENEIKON is in the process of moving this to a module.

File: spec/factories/authentications.rb (hunk 1 above)
Comment (OBCENEIKON) [inline spec/factories/authentications.rb]: Yes, this is for the authentications table. OmniAuth gives back a provider and uid for each provider.

File: app/models/staff.rb [hunk 3]
```diff
@@ -62,4 +63,20 @@ class Staff < ActiveRecord::Base
   def gravatar
     '3fc88b95c85e43f157cb1ffd0e37e832'
   end
+
+  def self.find_by_auth(auth_hash)
+    auth_match = Authentications.find_by(provider: auth_hash['proivider'], uid: auth_hash['uid'].to_s)
```
Comment (calebthompson) [inline app/models/staff.rb]: misspelled 'provider' here. Is there a test on this method?

File: config/initializers/omniauth.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+Rails.application.config.middleware.use OmniAuth::Builder do
+  # provider :developer unless Rails.env.production?
+  # provider :dropbox, ENV['DROPBOX_ACCESS_KEY'], ENV['DROPBOX_SECRET_KEY'], path_prefix: '/staff/auth'
+  # provider :ldap,
```
Comment (calebthompson) [inline config/initializers/omniauth.rb]: We're using LDAP - is any of this configured?

File: spec/factories/authentications.rb (hunk 1 above)
Comment (calebthompson) [inline spec/factories/authentications.rb]: From what I'm seeing, we'll want a `uid`, but it should be random, and the db should have a uniq index so we don't have multiple users signed in with a single facebook account.

Comment (calebthompson) [inline spec/factories/authentications.rb]: By "random" I mean something like `sequence(:uid)`, not necessarily random, just different between records.

Comment (OBCENEIKON) [inline spec/factories/authentications.rb]: UID comes from the provider, like a member ID.
https://github.com/intridea/omniauth/wiki/Auth-Hash-Schema

File: config/initializers/omniauth.rb (hunk 4 above)
Comment (OBCENEIKON) [inline config/initializers/omniauth.rb]: I was testing with it configured to jumpcloud

Comment (calebthompson) [inline config/initializers/omniauth.rb]: jellyfish-ldap probably should configure this (I assume that or similar is what you're building out).

**Comment by chriskacerguis** [comment]
Closing request after talking with @OBCENEIKON

**Pull request closed without merging**